### PR TITLE
Don't try to load previews when we know there is none

### LINKED ISSUE #559: Don't ask for a preview if we know we can't generate one.
### Steps to reproduce
1. Open the web developer console and switch to the network tab
1. Open a folder containing files with exotic file extensions (.lnk per example)

### Expected behaviour

Only files of a media type which is supported by the preview system (and enabled classes) should be allowed to ask for a preview

### Actual behaviour

Every files is asking for a preview, leading to a lot of wasted resources when the file is not of a supported media type.
Even with parallel request processing, it can lead to too many requests, which then slow down the experience, especially on low powered devices.

### Concept

1. Send a list of supported media types to JS 
1. We should only ask for a preview if the media type of the file is on that list

I thought we already had what we need in PHP, but I don't think we do.
In `OCP\IPreview` we can get the list of providers or ask if a media type is supported by the Preview system, which then caches the answer.

While looking for solutions, I stumbled upon the forgotten [`previewSupported`](
https://github.com/nextcloud/server/blob/8c7d7d7746e76b77ad86cee3aae5dbd4d1bcd896/apps/files_sharing/lib/Controllers/ShareController.php#L299) on the public share list, so maybe we could take the same approach and request for that information to be sent for each file which would make it easy to know when to ask for a preview.

----
@rullzer @longsleep

### PR DESCRIPTION
**Pull request by icewind1991**
Adds a dav property to tell whether or not a preview is available for a file. If the server tells us there is no preview for a file we don't try to load one

Fixes #559

To test:
- Upload some files with and without previews available
- Load the file list and check the network log for 404 errors when trying to load previews

### CODE REVIEW

**Comment by MorrisJobke** [comment]
```
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Files.FileList tests File selection Actions getSelectedFiles returns the selected file data FAILED
	Expected Object({ id: 1, name: 'One.txt', mimetype: 'text/plain', mtime: 123456789, type: 'file', size: 12, etag: 'abc', permissions: 31, hasPreview: true }) to equal Object({ id: 1, name: 'One.txt', mimetype: 'text/plain', mtime: 123456789, type: 'file', size: 12, etag: 'abc', permissions: 31 }).
	/drone/src/github.com/nextcloud/server/apps/files/tests/js/filelistSpec.js:1787:29
	Expected Object({ id: 3, name: 'Three.pdf', mimetype: 'application/pdf', mtime: 234560000, type: 'file', size: 58009, etag: '123', permissions: 31, hasPreview: true }) to equal Object({ id: 3, type: 'file', name: 'Three.pdf', mimetype: 'application/pdf', mtime: 234560000, size: 58009, etag: '123', permissions: 31 }).
	/drone/src/github.com/nextcloud/server/apps/files/tests/js/filelistSpec.js:1797:29
	Expected Object({ id: 4, name: 'somedir', mimetype: 'httpd/unix-directory', mtime: 134560000, type: 'dir', size: 250, etag: '456', permissions: 31, hasPreview: true }) to equal Object({ id: 4, type: 'dir', name: 'somedir', mimetype: 'httpd/unix-directory', mtime: 134560000, size: 250, etag: '456', permissions: 31 }).
	/drone/src/github.com/nextcloud/server/apps/files/tests/js/filelistSpec.js:1807:29
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Files.FileList tests File selection Actions Removing a file removes it from the selection FAILED
	Expected Object({ id: 1, name: 'One.txt', mimetype: 'text/plain', mtime: 123456789, type: 'file', size: 12, etag: 'abc', permissions: 31, hasPreview: true }) to equal Object({ id: 1, name: 'One.txt', mimetype: 'text/plain', mtime: 123456789, type: 'file', size: 12, etag: 'abc', permissions: 31 }).
	/drone/src/github.com/nextcloud/server/apps/files/tests/js/filelistSpec.js:1828:29
	Expected Object({ id: 4, name: 'somedir', mimetype: 'httpd/unix-directory', mtime: 134560000, type: 'dir', size: 250, etag: '456', permissions: 31, hasPreview: true }) to equal Object({ id: 4, type: 'dir', name: 'somedir', mimetype: 'httpd/unix-directory', mtime: 134560000, size: 250, etag: '456', permissions: 31 }).
	/drone/src/github.com/nextcloud/server/apps/files/tests/js/filelistSpec.js:1838:29
```

**Comment by icewind1991** [comment]
tests all green now

File: apps/dav/lib/Connector/Sabre/ServerFactory.php [hunk 1]
```diff
@@ -80,6 +83,7 @@ public function __construct(
 		$this->mountManager = $mountManager;
 		$this->tagManager = $tagManager;
 		$this->request = $request;
+		$this->previewManager = $previewManager;
```
Comment (MorrisJobke) [inline apps/dav/lib/Connector/Sabre/ServerFactory.php]: The property `previewManager` does not exist. Did you maybe forget to declare it?

**Comment by MorrisJobke** [comment]
Tested and it works 👍 

I can't judge on the webdav stuff that much. Maybe @rullzer can help here.

File: apps/dav/lib/Connector/Sabre/ServerFactory.php (hunk 1 above)
Comment (icewind1991) [inline apps/dav/lib/Connector/Sabre/ServerFactory.php]: fixed

**Comment by rullzer** [comment]
Awesome work!
LGTM

**Comment by MorrisJobke** [comment]
@karlitschek Port this to stable10?

**Comment by longsleep** [comment]
Would be awesome to get this as into a release as quickly as possible.

**Comment by karlitschek** [comment]
yes. please backport

**Comment by MorrisJobke** [comment]
@icewind1991 Could you open the backport PRs? Thanks

**Comment by icewind1991** [comment]
stable10: https://github.com/nextcloud/server/pull/610

**Pull request merged**